### ISSUE TITLE
[Cute, SM100] FA4 and FA4_Varlen fwd runtime difference reaches 27%

### ISSUE THREAD

**Issue by complexfilter** [Dao-AILab/flash-attention#2764]
version: 4.0.0b25
python version: 3.12
pytorch version: 2.12
cuda version: 13.2 
device: B200

(batch_size, seq_len, num_heads, dim) = (1, 86000, 32, 128)

```
numerics: dense and varlen outputs and gradients match.
dense  forward : 121.453 ms
varlen forward : 88.306 ms
dense  backward: 237.933 ms
varlen backward: 242.833 ms

forward  varlen/dense ratio: 0.73x
backward varlen/dense ratio: 1.02x
```

Reproduce:

```
"""Self-contained repro: Flash-Attention-4 dense vs varlen runtime parity.

For a single, fully-packed sequence the dense API (no cu_seqlens) and the varlen API
(cu_seqlens=[0, seq_len]) compute the exact same attention, so their forward and backward
kernels should run in ~the same time. This script benchmarks both and prints the ratio;
on our setup the two paths diverge, which is the bug we're reporting.

Requires a Hopper (sm90) or Blackwell (sm100) GPU and a flash-attn-4 build exposing
`flash_attn.cute.interface`. No other dependencies. Run: python <this file>.
"""

import torch
import triton
from flash_attn.cute.interface import _flash_attn_bwd, _flash_attn_fwd

BATCH_SIZE, SEQ_LEN, NUM_HEADS, HEAD_DIM = 1, 86000, 32, 128
SOFTMAX_SCALE = HEAD_DIM ** (-0.5)


def benchmark(f, name):
    ms = triton.testing.do_bench(f, warmup=20, rep=100)
    print(f"{name}: {ms:.3f} ms")
    return ms


def main():
    torch.manual_seed(42)
    shape = (BATCH_SIZE, SEQ_LEN, NUM_HEADS, HEAD_DIM)
    q = torch.randn(shape, dtype=torch.bfloat16, device="cuda")
    k = torch.randn(shape, dtype=torch.bfloat16, device="cuda")
    v = torch.randn(shape, dtype=torch.bfloat16, device="cuda")

    # Varlen inputs: drop the batch dim and describe the single segment with cu_seqlens.
    q_var, k_var, v_var = q[0], k[0], v[0]
    cu_seqlens = torch.tensor([0, SEQ_LEN], dtype=torch.int32, device="cuda")

    # flash-attn-4 >= 4.0.0b25 returns (out, lse, p, row_max); older builds return (out, lse).
    def dense_fwd():
        out, lse, *_ = _flash_attn_fwd(q, k, v, softmax_scale=SOFTMAX_SCALE, causal=False, return_lse=True)
        return out, lse

    def varlen_fwd():
        out, lse, *_ = _flash_attn_fwd(
            q_var, k_var, v_var,
            cu_seqlens_q=cu_seqlens, cu_seqlens_k=cu_seqlens,
            max_seqlen_q=SEQ_LEN, max_seqlen_k=SEQ_LEN,
            softmax_scale=SOFTMAX_SCALE, causal=False, return_lse=True,
        )
        return out, lse

    out, lse = dense_fwd()
    out_var, lse_var = varlen_fwd()
    dout = torch.randn_like(out)
    dout_var = dout[0]  # same upstream gradient, minus the batch dim, so the two bwd passes match

    def dense_bwd():
        return _flash_attn_bwd(
            q, k, v, out, dout, lse,
            softmax_scale=SOFTMAX_SCALE, causal=False,
            cu_seqlens_q=None, cu_seqlens_k=None, max_seqlen_q=None, max_seqlen_k=None,
        )

    def varlen_bwd():
        return _flash_attn_bwd(
            q_var, k_var, v_var, out_var, dout_var, lse_var,
            softmax_scale=SOFTMAX_SCALE, causal=False,
            cu_seqlens_q=cu_seqlens, cu_seqlens_k=cu_seqlens,
            max_seqlen_q=SEQ_LEN, max_seqlen_k=SEQ_LEN,
        )

    # The dense and varlen paths compute the same attention, so outputs and grads must match.
    # (Dense carries a leading batch dim; index it off before comparing to the varlen tensors.)
    torch.testing.assert_close(out[0], out_var, rtol=1e-3, atol=1e-3)
    torch.testing.assert_close(lse[0], lse_var, rtol=1e-3, atol=1e-3)
    dq, dk, dv = dense_bwd()
    dq_var, dk_var, dv_var = varlen_bwd()
    torch.testing.assert_close(dq[0], dq_var, rtol=1e-3, atol=1e-3)
    torch.testing.assert_close(dk[0], dk_var, rtol=1e-3, atol=1e-3)
    torch.testing.assert_close(dv[0], dv_var, rtol=1e-3, atol=1e-3)
    print("numerics: dense and varlen outputs and gradients match.")

    fwd_dense = benchmark(dense_fwd, "dense  forward ")
    fwd_varlen = benchmark(varlen_fwd, "varlen forward ")
    bwd_dense = benchmark(dense_bwd, "dense  backward")
    bwd_varlen = benchmark(varlen_bwd, "varlen backward")

    print(f"\nforward  varlen/dense ratio: {fwd_varlen / fwd_dense:.2f}x")
    print(f"backward varlen/dense ratio: {bwd_varlen / bwd_dense:.2f}x")


if __name__ == "__main__":
    main()
```
